src/librbd/image/RemoveRequest.cc

```diff
@@ -0,0 +1,641 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "common/dout.h"
+#include "common/errno.h"
+#include "common/WorkQueue.h"
+#include "common/ceph_context.h"
+#include "include/assert.h"
+#include "librbd/Utils.h"
+#include "librbd/internal.h"
+#include "librbd/ImageState.h"
+#include "librbd/ObjectMap.h"
+#include "librbd/ExclusiveLock.h"
+#include "librbd/MirroringWatcher.h"
+#include "journal/Settings.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "cls/journal/cls_journal_client.h"
+#include "librbd/image/RemoveRequest.h"
+#include "librbd/journal/RemoveRequest.h"
+#include "librbd/operation/TrimRequest.h"
+#include "librbd/mirror/DisableRequest.h"
+#include "librbd/operation/DisableFeaturesRequest.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::image::RemoveRequest: " << this << " " \
+                           << __func__ << " "
+
+namespace librbd {
+namespace image {
+
+using util::create_context_callback;
+using util::create_rados_ack_callback;
+
+template<typename I>
+RemoveRequest<I>::RemoveRequest(IoCtx &ioctx, const std::string &image_name, const std::string &image_id,
+                                bool force, ProgressContext &prog_ctx, ContextWQ *op_work_queue,
+                                Context *on_finish) :
+  m_image_name(image_name), m_image_id(image_id), m_force(force),
+  m_prog_ctx(prog_ctx), m_op_work_queue(op_work_queue), m_on_finish(on_finish) {
+  m_ioctx.dup(ioctx);
+  m_cct = reinterpret_cast<CephContext *>(m_ioctx.cct());
+
+  m_image_ctx = new I((m_image_id.empty() ? m_image_name : std::string()),
```

Comment:
This will be deleted in switch_thread_context. and this function is on the error path. it would be more clear in state diagram.